### PR TITLE
Host users, with different domain from server domain

### PR DESCRIPTION
**Pull request by rosbeef**
## Problem

- *No problem but improvement : Add ability to host user with different domain from server domain*

## Solution

- *Add well known host-meta file*
- *Add option to select users domain*

## PR Status

- [x] Code finished and ready to be reviewed/tested
  - [x] install
  - [x] upgrade
  - [x] remove
  - [x] restore
  - [x] change_url
- [ ] The fix/enhancement were manually tested (if applicable)
  - [x] 1rs install with master / host_domain: test.domain.tld
  - [x] update to patch1 
  - [x] change url / host_domain: social.domain.tld / user_domain: XXXXXXXXXX
  - [x] config userdomain / host_domain: XXXXXXXXXX / user_domain: social.domain.tld
  - [x] config userdomain / host_domain: XXXXXXXXXX / user_domain: domain.tld
  - [ ] remove app
  - [ ] restore app

## Automatic tests

!testme

Automatic tests can be triggered on https://ci-apps-dev.yunohost.org/ *after creating the PR*, by commenting "!testme", "!gogogadgetoci" or "By the power of systemd, I invoke The Great App CI to test this Pull Request!". (N.B. : for this to work you need to be a member of the Yunohost-Apps organization)

### CODE REVIEW

File: scripts/install [hunk 1]
```diff
@@ -122,6 +127,12 @@ chown -R $app:www-data "$final_path"
 #=================================================
 ynh_script_progression --message="Configuring NGINX web server..."
 
+# Create .well-known redirection for access by federation
+if yunohost --output-as plain domain list | grep -q "^$userdomain$"
```
Comment (Tagadda) [inline scripts/install:131]: I'm not sure why we check that... $userdomain should already be a valid domain, it's a `domain` in the manifest.
But we should check if we really need the redirection. If `domain` and `userdomain` are the same domain, we shouldn't add the file.
```suggestion
if [[ ! "$userdomain" == "$domain" ]]
```

File: scripts/install [hunk 2]
```diff
@@ -30,9 +30,13 @@ path_url="/"
...
 is_public=$YNH_APP_ARG_IS_PUBLIC
 language=$YNH_APP_ARG_LANGUAGE
-
+userdomain=$YNH_APP_ARG_USERDOMAIN
 admin_mail=$(ynh_user_get_info $admin 'mail')
 
+if [[ "$userdomain" == "$domain" ]]; then
+    userdomain=$domain    
+fi
+
```
Comment (Tagadda) [inline scripts/install:39]:
```suggestion
```
This doesn't do anything?

File: scripts/install [hunk 3]
```diff
@@ -242,6 +253,14 @@ ynh_add_config --template="../conf/cron" --destination="/etc/cron.d/$app"
 #=================================================
 # GENERIC FINALIZATION
 #=================================================
+# SETUP PERMISSIONS
+#=================================================
+#if yunohost --output-as plain domain list | grep -q "^$server_name$"; then
+#    ynh_permission_create --permission=server_client_infos --url=$userdomain/.well-known/host-meta \
+#                          --label="Server info for clients. (well-known)" --show_tile=false --allowed=visitors \
+#                          --auth_header=false --protected=true
+#fi
```
Comment (Tagadda) [inline scripts/install:262]:
```suggestion
if [[ ! "$userdomain" == "$domain" ]]; then
    ynh_permission_create --permission=server_client_infos --url=$userdomain/.well-known/host-meta \
                          --label="Server info for clients. (well-known)" --show_tile=false --allowed=visitors \
                          --auth_header=false --protected=true
fi
```
Something like that should work. But if `is_public` is false, the permission should not be protected, and should not allow visitors.
Maybe move this down in the `SETUP SSOWAT` section?

**Review by Tagadda — CHANGES_REQUESTED** [review]
Wow :open_mouth: This looks nice!
I suggested minor changes~

File: scripts/install (hunk 2 above)
Comment (rosbeef) [reply scripts/install:39]: seems not ;) sorry it's my first PR and not so good in dev ;)

File: scripts/install (hunk 3 above)
Comment (rosbeef) [reply scripts/install:262]: I used synapse package as example but i don know if it applies to mastodon.
i'm not so good with github. maybe it shoud not be in this PR

**Comment by rosbeef** [comment]
maybe it should be improved before publish.
i based my mod no https://docs.joinmastodon.org/dev/routes/

**Comment by rosbeef** [comment]
as it tell on https://dallas.lu/en/use-your-primary-domain-as-an-mastodon-name
fediverse servers like peertube which do not look for host-meta, should not work with main domain to subscribe to the video.
but subdomain should work

**Comment by rosbeef** [comment]
i don't find to make peertube subscription working with something like this in __HOSTDOMAIN__ :
location ~* ./well-known/webfinger {
    add_header Access-Control-Allow-Origin __USERDOMAIN__;
}

it should allow __HOSTDOMAIN__ to recieve crossSite request from __USERDOMAIN

File: conf/maindomain_nginx.conf [hunk 4]
```diff
@@ -0,0 +1,25 @@
+location = /.well-known/host-meta {
+ return 301 https://__DOMAIN__$request_uri;
+}
```
Comment (Tagadda) [inline conf/maindomain_nginx.conf:3]:
```suggestion
location ~ ^/.well-known/(webfinger|host-meta) {
 return 301 https://__DOMAIN__$request_uri;
}
```

File: conf/maindomain_nginx.conf [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+location ~ ^/(system|headers|avatars) {
+        ## set your mastodon public folder, or just redirect to $mastodon_host
+        rewrite ^(.*) https://__DOMAIN__$1 permanent;
+        #root /home/mastodon/live/public;
+}
+
+## mastodon url (possible use post)
+location ~ ^/(api/v1|inbox|actor|oauth|auth|users){
+        return 308 https://__DOMAIN__$request_uri;
+}
```
Comment (Tagadda) [inline conf/maindomain_nginx.conf:25]:
```suggestion
```

**Review by Tagadda — CHANGES_REQUESTED** [review]
All those redirections don't make much sense. We only need `/.well-known/host-meta` and `/.well-known/webfinger` on the `LOCAL_DOMAIN`.

I think you miss the `https` in front on the domain in `Access-Control-Allow-Origin`.

**Comment by yalh76** [comment]
!testme

**Comment by rosbeef** [comment]
i don't understand how to mitigate the Cross-Origin error ,
i put 
```
location ~* /.well-known/webfinger {
    add_header Access-Control-Allow-Origin https://__USERDOMAIN__;
}
```
in /etc/nginx/conf.d/__DOMAIN__.d/mastodon.conf
but peertube always return me a cors error

**Comment by Tagadda** [comment]
Maybe `Access-Control-Allow-Origin https://__USERDOMAIN__ https://__DOMAIN__;` ?

**Comment by rosbeef** [comment]
Appart script errors: i added parcial CORS mitigation
add:header Access-Control-Allow-Origin * to USERDOMAIN nginx conf

**Comment by yalh76** [comment]
!testme

**Comment by yalh76** [comment]
!testme

**Comment by rosbeef** [comment]
!testme

**Comment by rosbeef** [comment]
except for a bug that that i don't understand (mastodon can not find own served peertube on subscription from peertube with a mastodon account)
all is working for me.
And testing restore and upgrade most of time fail due to lake of space. @yalh76  have you an idea how to solve this problem ?

**Comment by yalh76** [comment]
!testme

**Comment by rosbeef** [comment]
!testme

**Comment by rosbeef** [comment]
@yalh76 !testme ;)

**Comment by ericgaspar** [comment]
!testme

**Comment by rosbeef** [comment]
i would take back my work on this pr but mixed up with push and saw that i was working on master...
are some one could help me to clean my pr and start back my workis it possible to revert my last forced-push action ?

formy information ; (To install Mastodon on mastodon.example.com in such a way it can serve @alice@example.com, set LOCAL_DOMAIN to example.com and WEB_DOMAIN to mastodon.example.com. This also requires additional configuration on the server hosting example.com to redirect or proxy requests to https://example.com/.well-known/webfinger to https://mastodon.example.com/.well-known/webfinger. For instance, with nginx, the configuration could look like the following:

location /.well-known/webfinger {
  add_header Access-Control-Allow-Origin '*';
  return 301 https://mastodon.example.com$request_uri;
}
) https://docs.joinmastodon.org/admin/config/#local_domain

**Comment by rosbeef** [comment]
Hi @ericgaspar you merged but it does noting. Some days ago I' be made a sync from source and all my modification were lost. I don't know how to go back.
Maybe it should be better to make an other PR but I would recover my last job to take it as example as i did it a long time ago.

**Pull request merged**